### PR TITLE
REF: Make read_json less stateful

### PR DESCRIPTION
**Pull request by mroeschke**
Also adds some validation for the `typ` keyword which aligns with the docs/typing

### CODE REVIEW

File: pandas/io/json/_json.py [hunk 1]
```diff
@@ -1292,42 +1286,31 @@ def _try_convert_to_date(self, data: Series) -> tuple[Series, bool]:
...
-    def _try_convert_types(self) -> None:
-        if self.obj is None:
-            return
-        obj, result = self._try_convert_data(
-            "data", self.obj, convert_dates=self.convert_dates
-        )
-        if result:
-            self.obj = obj
+    def _try_convert_types(self, obj: Series) -> Series:
+        obj, _ = self._try_convert_data("data", obj, convert_dates=self.convert_dates)
```
Comment (WillAyd) [inline pandas/io/json/_json.py:1301]: The existing code is pretty strange, but wouldn't this return something different in the case where the `result` value was False?

File: pandas/io/json/_json.py [hunk 2]
```diff
@@ -1383,48 +1366,43 @@ def _process_converter(
...
-                or col_lower == "modified"
-                or col_lower == "date"
-                or col_lower == "datetime"
-                or col_lower.startswith("timestamp")
-            ):
-                return True
-            return False
-
-        self._process_converter(lambda col, c: self._try_convert_to_date(c), filt=is_ok)
+            obj = self._process_converter(
```
Comment (WillAyd) [inline pandas/io/json/_json.py:1402]: Is it possible to make less stateless while maintaining these as separate functions? I find the new code harder to read, especially with the subtle difference in arguments to `self._process_converter`

File: pandas/tests/io/json/test_pandas.py [hunk 3]
```diff
@@ -792,7 +792,7 @@ def test_frame_from_json_precise_float(self):
 
     def test_typ(self):
         s = Series(range(6), index=["a", "b", "c", "d", "e", "f"], dtype="int64")
-        result = read_json(StringIO(s.to_json()), typ=None)
+        result = read_json(StringIO(s.to_json()), typ="series")
```
Comment (WillAyd) [inline pandas/tests/io/json/test_pandas.py:795]: Why did the tests need to change?

Comment (mroeschke) [reply pandas/tests/io/json/test_pandas.py:795]: This test didn't align with the docs/typing of `typ` accepting `None`. I added validation in this PR to ensure this case fails now

File: pandas/io/json/_json.py (hunk 2 above)
Comment (mroeschke) [reply pandas/io/json/_json.py]: Sure thing. I just pushed a change to remove `_process_converter` and make things simpler

File: pandas/io/json/_json.py (hunk 1 above)
Comment (mroeschke) [reply pandas/io/json/_json.py:1302]: I think `obj` _should_ be the same if result value was False i.e. the conversion was a no-op. I'm relying on that fact here that a no-op result should just be passed through here with `_try_convert_data`

File: pandas/tests/io/json/test_pandas.py (hunk 3 above)
Comment (WillAyd) [reply pandas/tests/io/json/test_pandas.py:795]: Hmm OK. That's super strange we allowed that to begin with...but hopefully harmless taking it away

File: pandas/io/json/_json.py [hunk 4]
```diff
@@ -985,16 +985,14 @@ def _get_object_parser(self, json) -> DataFrame | Series:
...
+            return FrameParser(json, **kwargs).parse()
+        elif typ == "series":
             if not isinstance(dtype, bool):
                 kwargs["dtype"] = dtype
-            obj = SeriesParser(json, **kwargs).parse()
-
-        return obj
+            return SeriesParser(json, **kwargs).parse()
+        else:
+            raise ValueError(f"{typ=} must be 'frame' or 'series'.")
```
Comment (WillAyd) [inline pandas/io/json/_json.py:995]: Maybe we should add a bugfix note for this at least to have it visibile. Something to the effect of "read_json now raises if the typ argument is neither frame nor series"

File: pandas/io/json/_json.py (hunk 1 above)
Comment (WillAyd) [reply pandas/io/json/_json.py:1302]: Ah OK - definitely not a fan of relying on the implementation of another function like that...but that can be cleaned up in a separate PR

**Review by WillAyd — APPROVED** [review]
nice to have a whatsnew note, but otherwise lgtm

**Pull request merged**